flax/linen/combinators.py

```diff
@@ -0,0 +1,44 @@
+"""Combinators of modules, such as a Sequential."""
+
+from typing import Callable, Sequence
+
+from flax.linen.module import Module
+
+class Sequential(Module):
+  """Applies a linear chain of Modules.
+
+  Meant to be used only for the simple case of fusing together callables where
+  the input of a particular module/op is the output of the previous one.
+
+  Modules will be applied in the order that they are passed in the constructor.
+
+  The apply() method of Sequential accepts any input and forwards it to the
+  first module it contains. It chains the output sequentially to the input of
+  the next module and returns the output of the final module.
+
+  Example usage:
+
+    class Foo(nn.Module):
+        feature_sizes: List[int]
+
+        @nn.compact
+        def __call__(self, x):
+        return nn.Sequential([nn.Dense(layer_size, name=f'layers_{idx}')
+                              for idx, layer_size
+                              in enumerate(self.feature_sizes)])(x)
+  """
+  layers: Sequence[Callable]
+
+  def __call__(self, inputs, *args, **kwargs):
+    outputs = inputs
+    for idx, layer in enumerate(self.layers):
+      if not callable(layer):
+        raise ValueError(
+            f"Each layer should implement `__call__`, but layer `{layer}` is "
+            "not callable."
+        )
+      if idx == 0:
+        outputs = layer(outputs, *args, **kwargs)
+      else:
+        outputs = layer(outputs)
+    return outputs
```

Comment:
Maybe we can get rid of `inputs` all together, since we can assume it will be contained in `*args`? Also, I wonder whether it is correct to return the input if the module is empty... maybe we should throw an error in that case? Otherwise it could potentially introduce bugs for the users (I don't think anyone would like to use an empty Sequential module).

How about something like this? We then don't explicitely check for `callable`, but that should be covered by the type checker. I believe the original PR had that check in as an attempt to stop the type checker to complain when using `Sequence[Module]`, but we are not using that anymore here.

```suggestion
  def __call__(self, *args, **kwargs):
    if not self.layers:
      raise ValueError(f'Empty Sequential module {self.name}.')

    y = self.layers[0](*args, **kwargs)
    for layer in self.layers[1:]:
      y = layer(y)
    return y
```

Reply:
Done.